api/src/main/java/net/starype/quiz/api/game/TimeOutEnding.java

```diff
@@ -0,0 +1,44 @@
+package net.starype.quiz.api.game;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+public class TimeOutEnding implements RoundEndingPredicate, Runnable {
+
+    private long time;
+    private TimeUnit unit;
+
+    private boolean isEnded;
+    private Runnable callBack;
+    private ScheduledExecutorService task;
+
+    public TimeOutEnding(long time, TimeUnit unit) {
+        this.unit = unit;
+        this.time = time;
+    }
+
+    public void startTimer(Runnable checkEndingCallback) {
+        this.callBack = checkEndingCallback;
+        this.task = Executors.newScheduledThreadPool(1);
+        task.schedule(this, time, unit);
+    }
+
+    @Override
+    public boolean ends() {
+        return isEnded;
+    }
+
+    @Override
+    public void run() {
+        this.isEnded = true;
+        callBack.run();

```

Comment:
Because we are using the Executors.newScheduledThreadPool we are creating a new thread here. So the callback must be a thread safe methods which hasn't been implemented for this project.

Reply:
It is impossible to implement a timer without using multithreading

Reply:
I have made suggestion in the comment section above. We can keep the idea of multithreading but this require to do it for likely all the methods of the API. We will also have to declare that in documentation as if someone want to implements his own callback version he will need to have a thread safe implementation

Reply:
I believe creating an event system here is the best way to go. At the worst scenario once we have that events system we could call it from another thread. But the point is it will be up to the user to decide so.

Reply:
Hi, I don't understand a shit about this but I just wanna say I'm a huge fan of @Tran-Antoine, so please don't hurt him

Reply:
[https://github.com/Tran-Antoine](@Tran-Antoine) is pretty pogchamp, please dont hit him too hard he already works a lot :/

Reply:
Yes, @Tran-Antoine is a very sensitive guy, so please be kind to him

Reply:
very interesting....